gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageProvider.java

```diff
@@ -0,0 +1,195 @@
+
+  /** Reverse map for storage reference to cache keys. */
+  final Map<Storage, StorageProviderCacheKey> storageToCacheKeyMap = new ConcurrentHashMap<>();
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  @VisibleForTesting
+  StorageProviderCacheKey computeCacheKey(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn) {
+    return StorageProviderCacheKey.builder()
+        .setCredentials(credentials)
+        .setHttpHeaders(storageOptions.getHttpRequestHeaders())
+        .setIsDirectPathPreferred(storageOptions.isDirectPathPreferred())
+        .setIsDownScopingEnabled(downscopedAccessTokenFn != null)
+        .setIsTracingEnabled(storageOptions.isTraceLogEnabled())
+        .setWriteChannelOptions(storageOptions.getWriteChannelOptions())
+        .setProjectId(storageOptions.getProjectId())
+        .build();
+  }
+
+  /** Determines if the storage instance can be served from the cache. */
+  private static boolean canCache(
+      GoogleCloudStorageOptions options,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService) {
+
+    return options.isStorageClientCachingExperimentEnabled()
+        // These values are currently passed while creating the storage client, but they are
+        // currently not configurable by the FS options so skipping caching.
+        && pCUExecutorService == null
+        && (interceptors == null || interceptors.isEmpty());
+  }
+
+  Storage getStorage(
+      Credentials credentials,
+      GoogleCloudStorageOptions storageOptions,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
+      throws IOException {
+    if (!canCache(storageOptions, interceptors, pCUExecutorService)) {
+      return createStorage(
+          credentials, storageOptions, interceptors, pCUExecutorService, downscopedAccessTokenFn);
+    }
+    StorageProviderCacheKey key =
+        computeCacheKey(credentials, storageOptions, downscopedAccessTokenFn);
+    Storage storage = cache.getIfPresent(key);
+    if (storage == null) {
+      storage = createStorage(credentials, storageOptions, null, null, downscopedAccessTokenFn);
+      cache.put(key, storage);
+      storageToCacheKeyMap.put(storage, key);
+      logger.atFinest().log(
+          "Cache miss, created new storage client. Cache hit count : %d, Cache hit rate : %d",
+          cache.stats().hitCount(), cache.stats().hitRate());
+    } else {
+      logger.atFinest().log(
+          "Cache hit, reusing the storage client. Cache hit count : %d, Cache hit rate : %d",
+          cache.stats().hitCount(), cache.stats().hitRate());
+    }
+    // Increment the reference count of the storage object.
+    storageClientToReferenceMap.put(
```

Comment:
We are performing 2 compound operations which needs to be atomic and thread safe:
1. Read through cache and update the counter.
2. Update the reverse hashmap with storage object to cache key mapping.

In current state the usage is not thread safe, we are manually reading using `cache.getIfPresent()` then doing a check over data and then manually writing `cache.put(key, storage)`. There is nothing avoiding multiple threads to read null from `cache.getIfPresent()` and write `cache.put(key, storage)` with counter value as 1. To make it thread safe, use the inbuilt methods of the ConcurrentHashMap which will internally take a lock or synchronize to avoid this scenario. 

```
StorageProviderCacheKey key = //compute cache key.
storageToCacheKeyMap.compute(key, (key, value) -> value == null ? createStorageAndUpdateReferenceMap() : increment(value));
```

Reply:
Thanks for catching this. I've made the compound operations atomic as opposed to individual atomic operations which may result in concurrency issues.